builtin/credential/appgroup/backend.go

```diff
@@ -0,0 +1,111 @@
+package appgroup
+
+import (
+	"sync"
+
+	"github.com/hashicorp/vault/helper/salt"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+type backend struct {
+	*framework.Backend
+	salt *salt.Salt
+
+	// Guards to clean-up the expired UserID entries
+	tidyUserIDCASGuard uint32
+
+	// Lock to make changes to registered Apps
+	appLock *sync.RWMutex
+
+	// Lock to make changes to registered Apps
+	groupLock *sync.RWMutex
+
+	// Lock to make changes to "generic" mode storage entries
+	genericLock *sync.RWMutex
+
+	// Map of locks to make changes to the UserIDs created.
+	// The lock in the map will be keyed off of UserID itself.
+	// Each UserID will have a separate lock which is used to
+	// update the information related to it, 'num_uses' for example.
+	// The lock will be deleted when the UserID is delted.
+	userIDLocks map[string]*sync.RWMutex
```

Comment:
This feels like it might be very memory expensive if someone has a whole lot of entries (keeping in mind that most HA backends can't scale to millions of keys anyways). You may want to look at what is being done in the `cubbyhole-the-world` branch to do something similar to token IDs, where we key it off of the first byte, giving 256 locks. That makes it easy to create them ahead of time at backend initialization, which in turns means you don't need to worry about map locking since you're only ever doing reads.

There's another subtle potential problem: it's not safe to have various goroutines reading and writing to a map at once. I don't know if this is also the case if you're having the map return default values, but if it does, you still need to then save that initialized value back in so that the next time someone does a read it's the same one. This is definitely true for a pointer, because the default value will be `nil` so unless you write the initialized value back in you won't be sharing mutexes.